### PR TITLE
Add "latest" flag to "yarn upgrade"

### PR DESCRIPTION
**Pull request by rally25rs**
Implements requested feature in #3384 

Adds a `yarn upgrade [--latest/-L]` flag.

This flag instructs Yarn to upgrade all dependencies to the latest version specified by it's registry, ignoring the versions specified in `package.json`.

This is functionally equivalent to running `yarn upgrade <package>` and passing all dependency names (`yarn upgrade foo bar baz ...etc`).

**Test plan**

In test file `__tests__/commands/upgrade.js` :
* Added additional tests for the existing `yarn upgrade` behaviors.
* Added tests for behavior specific to `latest` flag.

**Documentation**

PR for the website documentation repo should be merged at the same time:
https://github.com/yarnpkg/website/pull/513

### CODE REVIEW

**Comment by arcanis** [comment]
> This is functionally equivalent to running yarn upgrade <package> and passing all dependency names (yarn upgrade foo bar baz ...etc).

Shouldn't this flag be called `--all`, then?

**Comment by rally25rs** [comment]
I've always found this to be the most confusing part of Yarn;

I always expect `yarn upgrade` to be the same as `yarn upgrade <all packages>` but it isn't.

* If you specify a package name, it ignores the version range in package.json.
* If you don't specify a package name, then it uses the version ranges in package.json.

I think `--latest` is more indicative of what it's doing:

`yarn upgrade` - upgrades all respecting version range in package.json
`yarn upgrade --latest` - upgrades all to latest, ignoring package.json

In other words, `yarn upgrade` itself is already a "--all" in that it upgrades all packages. This change is really about what version it upgrades all packages to.

... hope that makes sense. It's early and I'm on my 1st cup ☕️

**Comment by cpojer** [comment]
I like `--latest`, I think the explanation makes sense.

File: src/cli/commands/upgrade.js [hunk 1]
```diff
@@ -50,6 +52,11 @@ export async function run(config: Config, reporter: Reporter, flags: Object, arg
     } else {
       throw new MessageError(reporter.lang('scopeNotValid'));
     }
+  } else if (flags.latest && args.length === 0) {
+    addArgs = Object.keys(allDependencies)
+      .map(dependency => {
+        return getDependency(allDependencies, dependency);
+      });
```
Comment (cpojer) [inline src/cli/commands/upgrade.js]: nit: you don't need the curly braces and the return here.

File: src/cli/commands/upgrade.js [hunk 2]
```diff
@@ -11,6 +11,7 @@ export function setFlags(commander: Object) {
   // TODO: support some flags that install command has
   commander.usage('upgrade [flags]');
   commander.option('-S, --scope <scope>', 'upgrade packages under the specified scope');
+  commander.option('-L, --latest', 'upgrade packages to the latest version, ignoring version ranges in package.json');
```
Comment (cpojer) [inline src/cli/commands/upgrade.js]: do we really need the `-L` shorthand? Let's just go with `--latest`.

**Comment by cpojer** [comment]
Would you mind rebasing this diff and making it pass on CI? :)

File: src/cli/commands/upgrade.js (hunk 2 above)
Comment (rally25rs) [inline src/cli/commands/upgrade.js]: OK, was just following the convention of the other flags. I'll remove that `-L`.

**Comment by rally25rs** [comment]
oops, I made the classic mistake of `git commit -a` and forgetting to `git add` new files 😢 
Merged latest master into the branch and fixed tests. Should be good to go now @cpojer

File: __tests__/commands/upgrade.js [hunk 3]
```diff
@@ -102,6 +102,70 @@ test.concurrent('upgrades from fixed version to latest', (): Promise<void> => {
...
+test.concurrent('--latest upgrades to latest ignoring package.json when no package name passed', (): Promise<void> => {
+  return runUpgrade([], {latest: true}, 'range-to-latest', async (config): ?Promise<void> => {
+    const lockfile = explodeLockfile(await fs.readFile(path.join(config.cwd, 'yarn.lock')));
+    const pkg = await fs.readJson(path.join(config.cwd, 'package.json'));
+    const lockEntryIndex = lockfile.indexOf('left-pad@^1.1.3:');
+
+    expect(lockEntryIndex).toEqual(0);
+    expect(lockfile.indexOf('left-pad@<=1.1.1:')).toEqual(-1);
+    expect(lockfile[lockEntryIndex + 1]).toContain('1.1.3');
+    expect(pkg.dependencies).toEqual({'left-pad': '^1.1.3'});
```
Comment (arcanis) [inline __tests__/commands/upgrade.js]: You shouldn't use `left-pad` for this test, otherwise it will break if they release a new version. I think you can use `yarn-test` (cc @kittens ?)

File: __tests__/commands/upgrade.js [hunk 4]
```diff
@@ -102,6 +102,70 @@ test.concurrent('upgrades from fixed version to latest', (): Promise<void> => {
...
+
+test.concurrent('upgrades to latest matching semver when package name passed with version', (): Promise<void> => {
+  return runUpgrade(['left-pad@~1.1.2'], {}, 'range-to-latest', async (config): ?Promise<void> => {
+    const lockfile = explodeLockfile(await fs.readFile(path.join(config.cwd, 'yarn.lock')));
+    const pkg = await fs.readJson(path.join(config.cwd, 'package.json'));
+    const lockEntryIndex = lockfile.indexOf('left-pad@~1.1.2:');
+
+    expect(lockEntryIndex).toEqual(0);
+    expect(lockfile[lockEntryIndex + 1]).toContain('1.1.3');
+    expect(pkg.dependencies).toEqual({'left-pad': '~1.1.2'});
```
Comment (arcanis) [inline __tests__/commands/upgrade.js]: You can use `left-pad` here, but then use `~1.0.1` (should resolve to `1.0.2`) instead of `~1.1.2`, since we know there won't be any more release to this branch, which is not the case for `1.1.x`.

File: __tests__/commands/upgrade.js [hunk 5]
```diff
@@ -102,6 +102,70 @@ test.concurrent('upgrades from fixed version to latest', (): Promise<void> => {
...
+  return runUpgrade(['left-pad'], {}, 'range-to-latest', async (config): ?Promise<void> => {
+    const lockfile = explodeLockfile(await fs.readFile(path.join(config.cwd, 'yarn.lock')));
+    const pkg = await fs.readJson(path.join(config.cwd, 'package.json'));
+    const lockEntryIndex = lockfile.indexOf('left-pad@^1.1.3:');
+
+    expect(lockEntryIndex).toEqual(0);
+    expect(lockfile.indexOf('left-pad@<=1.1.1:')).toEqual(-1);
+    expect(lockfile[lockEntryIndex + 1]).toContain('1.1.3');
+    expect(pkg.dependencies).toEqual({'left-pad': '^1.1.3'});
+  });
```
Comment (arcanis) [inline __tests__/commands/upgrade.js]: This doesn't feel right ... shouldn't we have to pass the `--latest` flag to have this behavior? It contradicts the "no argument means all packages" behavior ("upgrades to latest matching package.json semver when no package name passed").

**Comment by cpojer** [comment]
Oops, that was within the same 15 seconds! @rally25rs would you mind sending a follow-up PR to fix the issues that @arcanis pointed out?

File: __tests__/commands/upgrade.js (hunk 5 above)
Comment (rally25rs) [inline __tests__/commands/upgrade.js]: This test actually passes in current Yarn without my modifications (in other words, this is how Yarn has worked for a while). This is testing the command `yarn upgrade left-pad`, which is roughly equivalent to `yarn add left-pad`. The `add` command resolves to the passed in version range for `left-pad` which is unspecified, so it resolves to latest. 

There is some discussion around that confusing behavior in 3384 3266 3204
I've always sort of wondered if it's an unintentional behavior.

File: __tests__/commands/upgrade.js (hunk 3 above)
Comment (rally25rs) [inline __tests__/commands/upgrade.js]: Hmm, I actually thought the packages used during tests were all loaded from the 'offline mirror' directory of `__tests__/fixtures/request-cache/...` but it does like like just the left-pad tgz files are in there, not the metadata. So all these tests are actually resolving real metadata? that seems fragile...

I needed a package that had enough diverse version numbers to cover all these upgrade cases, and left-pad was the first one I found that worked.

Is there someplace I can see what version of `yarn-test` exist?

(side-note, it might be useful to document how these tests actually work and should be constructed, for new developers who want to contribute. It took me a while the first time I needed to add one.)

File: __tests__/commands/upgrade.js (hunk 5 above)
Comment (rally25rs) [inline __tests__/commands/upgrade.js]: I just re-read my comment and decided it's confusing... In clearer terms, the way Yarn 0.24 - 0.26 all work currently is that if your package.json is:

```
"dependencies": {
  "left-pad": "^1.0.0"
```

Then `yarn upgrade` will upgrade left-pad to the greatest match for `^1.0.0`.
However `yarn upgrade left-pad` will upgrade to `latest`.

I didn't change that behavior.
This PR just adds the option to:

`yarn upgrade --latest` which will upgrade left-pad to `latest` (ignoring what was in package.json)

There is currently, and with this PR, still no way to `yarn upgrade <package_name>` and have it use the range specified in package.json, unless you manually copy-paste the range yourself and do `yarn upgrade left-pad@^1.0.0`

Comment (arcanis) [inline __tests__/commands/upgrade.js]: Yep, you're right. I feel like this should be considered as a bug ... otherwise there's no way to upgrade a specific package inside its allowed boundaries :/ can you open a separate PR where we could discuss this?

File: __tests__/commands/upgrade.js (hunk 3 above)
Comment (arcanis) [inline __tests__/commands/upgrade.js]: > So all these tests are actually resolving real metadata? that seems fragile...

Think so. That's why our tests practically never completely pass: we always end up timeouting at least one of them 😞 Probably should fix this at some point! I thought maybe we could set up some sort of fake registry with only the few endpoints required, that would extract its data from a repository local folder.

File: __tests__/commands/upgrade.js (hunk 5 above)
Comment (rally25rs) [inline __tests__/commands/upgrade.js]: I spawned #3603 as a discussion issue and added you as an assignee.

File: __tests__/commands/upgrade.js (hunk 3 above)
Comment (rally25rs) [inline __tests__/commands/upgrade.js]: We can discuss this in #3455

**Comment by develar** [comment]
One problem with this flag — range specifier is not preserved:

 `"electron-osx-sign": "0.4.6"` 

changed to 

`"electron-osx-sign": "^0.4.6"`

i.e. package not upgraded, but range specifier changed.

Range specifier must be preserved.

**Comment by arcanis** [comment]
@develar good remark, can you open an issue?

**Comment by develar** [comment]
Filed https://github.com/yarnpkg/yarn/issues/3609 .

File: __tests__/commands/upgrade.js (hunk 3 above)
Comment (rally25rs) [inline __tests__/commands/upgrade.js]: The metadata for `left-pad` is taken from the `__tests__/fixtures/request-cache/` directory, so a new version pushed to the live registry would not break the test. These tests actually pass with no internet connection.
